ui/store/action-queue/index.js

```diff
@@ -0,0 +1,200 @@
+//   }, {});
+// }
+
+let background = null;
+let promisifiedBackground = null;
+
+const actionRetryQueue = [];
+
+function failQueue() {
+  actionRetryQueue.forEach(({ reject }) =>
+    reject(
+      Error('Background operation cancelled while waiting for connection.'),
+    ),
+  );
+}
+
+/**
+ * Drops the entire actions queue. Rejects all actions in the queue unless silently==true
+ * Does not affect the single action that is currently being processed.
+ *
+ * @param {boolean} [silently]
+ */
+export function dropQueue(silently) {
+  if (!silently) {
+    failQueue();
+  }
+  actionRetryQueue.length = 0;
+}
+
+// add action to queue
+const executeActionOrAddtoRetryQueue = (item) => {
+  if (actionRetryQueue.some((act) => act.actionId === item.actionId)) {
+    return;
+  }
+
+  if (background.connectionStream.readable) {
+    executeAction({
+      action: item,
+      disconnectSideeffect: () => actionRetryQueue.push(item),
+    });
+  } else {
+    actionRetryQueue.push(item);
+  }
+};
+
+/**
+ * Promise-style call to background method
+ * In MV2: invokes promisifiedBackground method directly.
+ * In MV3: action is added to retry queue, along with resolve handler to be executed on completion,
+ *  the queue is then immediately processed if background connection is available.
+ *  On completion (successful or error) the action is removed from the retry queue.
+ *
+ * @param {string} method - name of the background method
+ * @param {Array} [args] - arguments to that method, if any
+ * @param {any} [actionId] - if an action with the === same id is submitted, it'll be ignored if already in queue waiting for a retry.
+ * @returns {Promise}
+ */
+export function submitRequestToBackground(
+  method,
+  args = [],
+  actionId = Date.now() + Math.random(), // current date is not guaranteed to be unique
+) {
+  if (isManifestV3) {
+    return new Promise((resolve, reject) => {
+      executeActionOrAddtoRetryQueue({
+        actionId,
+        request: { method, args },
+        resolve,
+        reject,
+      });
+    });
+  }
+  return promisifiedBackground[method](...args);
+}
+
+/**
+ * Callback-style call to background method
+ * In MV2: invokes promisifiedBackground method directly.
+ * In MV3: action is added to retry queue, along with resolve handler to be executed on completion,
+ *  the queue is then immediately processed if background connection is available.
+ *  On completion (successful or error) the action is removed from the retry queue.
+ *
+ * @param {string} method - name of the background method
+ * @param {Array} [args] - arguments to that method, if any
+ * @param callback - Node style (error, result) callback for finishing the operation
+ * @param {any} [actionId] - if an action with the === same id is submitted, it'll be ignored if already in queue.
+ */
+export const callBackgroundMethod = (
+  method,
+  args = [],
+  callback,
+  actionId = Date.now() + Math.random(), // current date is not guaranteed to be unique
```

Comment:
I think I'm missing where we use this 4th parameter. Aside from the initial background calls, is it used elsewhere?

Reply:
it is used to check if action is already present in queue

Reply:
may I have an example of what file and what line this is being called in please? 🙏

Reply:
I'm not aware of it being used (as in calling the function with the 4th argument). We're currently only using the default value generated unique each time. But the capability is there to ensure only one copy of a specific type of action in retry queue. 

That being said, it's been preserved as a requirement from before I started working on this and I don't have an explanation why it was needed. Might no longer make sense. :shrug:

Reply:
Please check these PRs @digiwand :
https://github.com/MetaMask/metamask-extension/pull/15667
https://github.com/MetaMask/metamask-extension/pull/15675

Reply:
I think the intention of having it makes sense to ensure we're not duplicating calls of an action instance in the off chance an action could fail. I'm wondering, why add it as a parameter instead of using a const in the function? 

The examples above look like they're only for `submitRequestToBackground` and not `callBackgroundMethod`

Reply:
I think the intention of having it makes sense to ensure we're not duplicating retried action calls in the off chance an action could fail. I'm wondering, why add it as a parameter instead of using a const in the function? 

I'm not seeing where `callBackgroundMethod` gets called with the 4th parameter in either #15667 or #15675. The examples above look like they're only for `submitRequestToBackground` and not `callBackgroundMethod`

Reply:
the callback-style syntax can hopefully go away over time.